potassium/potassium.py

```diff
@@ -1,70 +1,84 @@
+                self._mailbox[request_id] = ThreadQueue()
+        result, stream_id = self._mailbox[request_id].get()
+
+        if stream_id is not None:
+            result.body = self._stream_body(stream_id)
+
+        with self._lock:
+            del self._mailbox[request_id]
+
+        return result
+
+    def _stream_body(self, stream_id):
+        with self._lock:
+            if stream_id not in self._mailbox:
+                self._mailbox[stream_id] = ThreadQueue()
+        queue = self._mailbox[stream_id]
 
-
-class Endpoint():
-    def __init__(self, type, func):
-        self.type = type
-        self.func = func
-
-class Request():
-    def __init__(self, id: str, headers: EnvironHeaders, json: dict):
-        self.id = id
-        self.headers = headers
-        self.json = json
-
-ResponseBody = Union[bytes, Generator[bytes, None, None]]
-
-class Response():
-    def __init__(self, status: int = 200, json: Optional[dict] = None, headers: Optional[dict] = None, body: Optional[ResponseBody] = None):
-        assert json == None or body == None, "Potassium Response object cannot have both json and body set"
-
-
-        self.headers = headers if headers != None else {}
-
-        # convert json to body if not None
-        if json != None:
-            self.body = jsonlib.dumps(json).encode("utf-8")
-            self.headers["Content-Type"] = "application/json"
-        else:
-            self.body = body
-
-        self.status = status
-
-    @property
-    def json(self):
-        if self.body == None:
-            return None
-        if type(self.body) == bytes:
-            try:
-                return jsonlib.loads(self.body.decode("utf-8"))
-            except:
-                return None
-        return None
-            
-    @json.setter
-    def json(self, json):
-        self.body = jsonlib.dumps(json).encode("utf-8")
-        self.headers["Content-Type"] = "application/json"
-
-
-class InvalidEndpointTypeException(Exception):
-    def __init__(self):
-        super().__init__("Invalid endpoint type. Must be 'handler' or 'background'")
-
-
-class RouteAlreadyInUseException(Exception):
-    def __init__(self):
-        super().__init__("Route already in use")
+        try:
+            while True:
+                result = queue.get()
+                if isinstance(result, Exception):
+                    with self._lock:
+                        del self._mailbox[stream_id]
+                    raise result
+                elif result == None:
+                    break
+                else:
+                    yield result
+        except GeneratorExit:
+            while True:
+                # flush the queue
+                result = queue.get()
+                if result == None:
+                    break
+                elif isinstance(result, Exception):
+                    with self._lock:
+                        del self._mailbox[stream_id]
+                    raise result
```

Comment:
this could probably be extracted into some private methods, cognitive complexity is fairly high.